### PR TITLE
*: handle truncate error 

### PR DESCRIPTION
**Pull request by overvenus**
Handle truncate error.

For now the error message sent by tikv is different from tidb's. 

TiDB:

    ERROR 1105 (HY000): [1 [types:1265]Data Truncated

TiKV:

    ERROR 1105 (HY000): other error: unknown error Codec(Other(StringError("[1265] Data Truncated")))

PTAL @BusyJay @coocood @siddontang 

CC pingcap/tidb#2212

### CODE REVIEW

File: src/util/codec/convert.rs [hunk 1]
```diff
@@ -112,4 +177,53 @@ mod test {
...
+                        },
+                        EvalContext {
+                            tz: FixedOffset::east(0),
+                            ignore_truncate: false,
+                            truncate_as_warning: false,
+                        }];
+
+        for ctx in &ctxs {
+            assert!(super::handle_truncate(ctx, false).is_ok());
+            assert!(super::handle_truncate(ctx, false).is_ok());
```
Comment (coocood) [inline src/util/codec/convert.rs]: Duplicated code?

**Comment by coocood** [comment]
LGTM

File: src/util/codec/convert.rs [hunk 2]
```diff
@@ -61,11 +63,74 @@ pub fn bytes_to_f64(bytes: &[u8]) -> Result<f64> {
     Ok(f)
 }
 
+/// `bytes_to_f64` converts a byte array to a float64 in best effort.
+/// similar to `bytes_to_f64` with additional context.
+pub fn bytes_to_f64_with_context(ctx: &EvalContext, bytes: &[u8]) -> Result<f64> {
```
Comment (BusyJay) [inline src/util/codec/convert.rs]: refactor `bytes_to_f64` instead.

File: src/util/codec/convert.rs [hunk 3]
```diff
@@ -41,8 +39,17 @@ pub fn bytes_to_int(bytes: &[u8]) -> Result<i64> {
     Ok(r)
 }
 
-/// `bytes_to_f64` converts a byte array to a float64 in best effort.
-pub fn bytes_to_f64(bytes: &[u8]) -> Result<f64> {
+/// `bytes_to_int` converts a byte arrays to an i64 in best effort.
+/// TODO: handle overflow.
+pub fn bytes_to_int(ctx: &EvalContext, bytes: &[u8]) -> Result<i64> {
+    let s = try!(str::from_utf8(bytes)).trim();
+    let vs = get_valid_float_prefix(s);
```
Comment (BusyJay) [inline src/util/codec/convert.rs]: Why int value needs a valid float prefix?

File: src/util/codec/mysql/decimal.rs [hunk 4]
```diff
@@ -1427,7 +1426,10 @@ impl Decimal {
         }
 
         if end_idx + 1 < bs.len() && (bs[end_idx] == b'e' || bs[end_idx] == b'E') {
-            let exp = try!(convert::bytes_to_int(&bs[end_idx + 1..]));
+            let exp = {
```
Comment (BusyJay) [inline src/util/codec/mysql/decimal.rs]: Why change this?

File: src/util/codec/convert.rs (hunk 3 above)
Comment (overvenus) [inline src/util/codec/convert.rs]: Just follows TiDB https://github.com/pingcap/tidb/blob/bb66fbe6b0ff7c09057d8b6821cec6c4832e2628/util/types/convert.go#L138.
However, overflow is not handled yet.

**Pull request merged**